cli/cli.ts

```diff
@@ -3367,6 +3377,84 @@ export function extractAsync(filename: string) {
         })
 }
 
+export function preCacheHexAsync() {
+    let trgInfo = readLocalPxTarget();
+
+    if (!trgInfo) {
+        console.log("pxtarget.json not found; make sure you are in a valid PXT target directory");
+        return Promise.resolve();
+    }
+
+    // Extract the target's default project to disk
+    if (!trgInfo.blocksprj) {
+        console.log("Could not find default project 'blocksprj' in pxtarget.json");
+        return Promise.resolve();
+    }
+
+    let projectsRoot = "projects";
+    let projectInitialName = "_precache_hex";
+    let projectName = projectInitialName;
+    let projectPath = path.join(projectsRoot, projectName);
+    let counter = 2;
+
+    while (fs.existsSync(projectPath)) {
+        projectName = projectInitialName + "-" + counter;
+        projectPath = path.join(projectsRoot, projectName);
+        ++counter;
+    }
+
+    nodeutil.mkdirP(projectPath);
+    trgInfo.blocksprj.config.name = projectName;
+    fs.writeFileSync(path.join(projectPath, "pxt.json"), JSON.stringify(trgInfo.blocksprj.config, null, 4));
+    Object.keys(trgInfo.blocksprj.files).forEach((f) => {
+        fs.writeFileSync(path.join(projectPath, f), trgInfo.blocksprj.files[f]);
+    });
+    console.log("default project extracted");
+
+    // Install package dependencies
+    let previousCwd = process.cwd();
+    process.chdir(projectPath);
+
+    return installAsync()
+        .then(() => {
+            console.log("packages installed");
+
+            // Build in the cloud
+            forceCloudBuild = true;
+            return buildAsync();
+        })
+        .then((compileOpts: pxtc.CompileOptions) => {
+            if (!compileOpts) {
+                console.log("Failed to extract HEX image");
+                return;
+            }
+
+            // Place the base HEX image in the hex cache if necessary
+            let sha = compileOpts.extinfo.sha;
+            let hex: string[] = compileOpts.hexinfo.hex;
+            let hexCache = path.join(previousCwd, "hexcache");
+            let hexFile = path.join(hexCache, sha + ".hex");
+
+            process.chdir(previousCwd);
+            nodeutil.mkdirP(hexCache);
+
+            if (fs.existsSync(hexFile)) {
+                console.log("HEX image already in offline cache");
+            } else {
+                fs.writeFileSync(hexFile, hex.join(os.EOL));
+                console.log(`Created HEX image in offline cache: ${hexFile}`);
+            }
+
+            // Clean up temp project
+            nodeutil.deleteFolderRecursive(projectPath);
```

Comment:
we can leave the folder as is as long as we nuke it when rebuilding. sometimes it's helpful to inspect files on disk.